### PR TITLE
Add first iteration of developer wallet

### PR DESCRIPTION
**Pull request by deangalvin-cb**
### What changed? Why?
This adds our developer custodial model for staking to the developer wallet.

#### Qualified Impact
This adds the ability for users to stake, sign, and broadcast in a single function. This creates a staking operation and tracks transaciton status through the lifecycle.

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: lib/coinbase/address/wallet_address.rb [hunk 1]
```diff
@@ -215,5 +245,44 @@ def broadcast_trade(trade, signed_payload:, approve_tx_signed_payload: nil)
...
+      end
+
+      StakingOperation.new(operation_model)
+    end
+
+    def complete_staking_operation(amount, asset_id, action, mode: :default, options: {})
+      validate_staking_operation!(amount, asset_id, action, mode: mode, options: options)
+      op = create_staking_operation(amount, asset_id, action, mode: mode, options: options)
+      op.transactions.each_with_index do |transaction, i|
+        signed_payload = transaction.sign(@key)
```
Comment (marcin-cb) [inline lib/coinbase/address/wallet_address.rb:275]: To avoid any extra errors would be good to broadcast only tx's that are signed and check that with the `signed?` function on the `Transaction` object.

File: lib/coinbase/address/wallet_address.rb [hunk 2]
```diff
@@ -215,5 +245,44 @@ def broadcast_trade(trade, signed_payload:, approve_tx_signed_payload: nil)
...
+        Coinbase.call_api do
```
Comment (marcin-cb) [inline lib/coinbase/address/wallet_address.rb:276]: Don't we want to poll and wait for the status to be complete with `getStakingOperation`?

Comment (deangalvin-cb) [reply lib/coinbase/address/wallet_address.rb:276]: Ah, like all of the tx's generated?

Comment (marcin-cb) [reply lib/coinbase/address/wallet_address.rb:276]: Yeah. This would work for PES but with multiple tx's, each tx would have to broadcast successfully for the next one to continue. 
Also even with a single transaction, we'd like to poll the status for staking operation and update the status on the SDK side to reflect the status in the backend.

Comment (deangalvin-cb) [reply lib/coinbase/address/wallet_address.rb]: That should happen in the wait function. Lets regroup on this for >32 eth as I believe a different paradigm is needed.

File: spec/unit/coinbase/staking_operation_spec.rb [hunk 3]
```diff
@@ -86,4 +168,31 @@
       end
     end
   end
+
+  describe '#broadcast!' do
```
Comment (alex-stone) [inline spec/unit/coinbase/staking_operation_spec.rb:172]: Can we test the multi-transaction scenario?

File: lib/coinbase/address.rb [hunk 4]
```diff
@@ -75,10 +75,138 @@ def faucet
       end
     end
 
+    # Builds a stake operation for the supplied asset. The stake operation
```
Comment (drohit-cb) [inline lib/coinbase/address.rb:78]: @marcin-cb this is a nice clarification we can add to nodejs as well

File: lib/coinbase/address.rb [hunk 5]
```diff
@@ -75,10 +75,138 @@ def faucet
...
+    # @param mode [Symbol] The staking mode. Defaults to :default.
+    # @param options [Hash] Additional options for the unstake operation
+    # @return [Coinbase::StakingOperation] The unstake operation
+    def build_unstake_operation(amount, asset_id, mode: :default, options: {})
+      validate_can_perform_staking_action!(amount, asset_id, 'unstakeable_balance', mode, options)
+
+      StakingOperation.build(amount, network_id, asset_id, id, 'unstake', mode, options)
+    end
+
+    # Builds an claim_stake operation for the supplied asset.
```
Comment (drohit-cb) [inline lib/coinbase/address.rb:103]:
```suggestion
    # Builds a claim_stake operation for the supplied asset.
```

File: lib/coinbase/address.rb [hunk 6]
```diff
@@ -75,10 +75,138 @@ def faucet
...
+      staking_balances(asset_id, mode: mode, options: options)[:claimable_balance]
+    end
+
+    # Lists the staking rewards for the address.
+    # @param asset_id [Symbol] The asset to retrieve staking rewards for
+    # @param start_time [Time] The start time for the rewards. Defaults to 1 month ago.
+    # @param end_time [Time] The end time for the rewards. Defaults to the current time.
+    # @param format [Symbol] The format to return the rewards in. Defaults to :usd.
+    # @return [Enumerable<Coinbase::StakingReward>] The staking rewards
+    def staking_rewards(asset_id, start_time: DateTime.now.prev_month(1), end_time: DateTime.now, format: :usd)
```
Comment (drohit-cb) [inline lib/coinbase/address.rb:186]: maybe use the same lookback default of nodejs i.e. a week ? https://github.com/coinbase/coinbase-sdk-nodejs/blob/e40237144507247c82055e268f97e2b0b7cf97c0/src/coinbase/address.ts#L93

File: lib/coinbase/address/wallet_address.rb [hunk 7]
```diff
@@ -144,5 +177,13 @@ def ensure_sufficient_balance!(amount, asset_id)
 
       raise InsufficientFundsError.new(amount, current_balance)
     end
+
+    def complete_staking_operation(amount, asset_id, action, mode: :default, options: {})
+      op = StakingOperation.create(amount, network_id, asset_id, id, wallet_id, action, mode, options)
+      op.transactions.each do |transaction|
+        transaction.sign(@key)
```
Comment (drohit-cb) [inline lib/coinbase/address/wallet_address.rb:184]: nodejs has checks around skipping signing if already signed - is that something we should do here as well ?

File: lib/coinbase/address/wallet_address.rb [hunk 8]
```diff
@@ -81,9 +81,6 @@ def trade(amount, from_asset_id, to_asset_id)
         wallet_id: wallet_id
       )
 
-      # If a server signer is managing keys, it will sign and broadcast the underlying trade transaction out of band.
```
Comment (drohit-cb) [inline lib/coinbase/address/wallet_address.rb:84]: is this related to this PR ?

File: lib/coinbase/address/wallet_address.rb (hunk 7 above)
Comment (deangalvin-cb) [reply lib/coinbase/address/wallet_address.rb:184]: No as we dont do it for other actions

File: lib/coinbase/address.rb (hunk 6 above)
Comment (deangalvin-cb) [reply lib/coinbase/address.rb:186]: Not sure when it changed for NodeJS, but this has always been a month, we also use 10 days everywhere in our docs, we should better align

File: lib/coinbase/address/wallet_address.rb (hunk 8 above)
Comment (deangalvin-cb) [reply lib/coinbase/address/wallet_address.rb:84]: Nah this is just a git relic, you can ignore.

**Review by drohit-cb — APPROVED** [review]

File: lib/coinbase/address.rb (hunk 6 above)
Comment (drohit-cb) [reply lib/coinbase/address.rb]: yeah I think it will be nice to update the docs to remove the code around last 10 days and have it be defaulted to whatever we have here.
